File: _legacy/python/cwlint/engine.py
"""Run all registered rules against a target repo."""

from __future__ import annotations

from typing import TYPE_CHECKING

from cwlint.discovery import discover
from cwlint.findings import Finding, Report
from cwlint.rules import all_rules

if TYPE_CHECKING:
    from collections.abc import Iterable
    from pathlib import Path

    from cwlint.spec import Spec


def check_repo(root: Path, spec: Spec, *, ignore: Iterable[str] = ()) -> Report:
    """Discover and run all rules against `root`. Read-only."""
    layout = discover(root)
    ignored = set(ignore)
    findings: list[Finding] = []
    for rule in all_rules():
        if rule.rule_id in ignored:
            continue
        findings.extend(rule.check(layout, spec))
    return Report(
        spec_version=spec.spec_version,
        claude_app_version=spec.claude_app_version,
        findings=sorted(findings, key=lambda f: (f.rule_id, str(f.path), f.line)),
    )

File: src/engine.ts
/**
 * Run all rules against a target repo and produce a Report.
 * Mirrors `cwlint.engine.check_repo`.
 */

import { discover } from "./discovery.js";
import type { Finding, Report } from "./findings.js";
import { ALL_RULES } from "./rules.js";
import type { Spec } from "./spec.js";

export function checkRepo(root: string, spec: Spec, opts: { ignore?: string[] } = {}): Report {
  const layout = discover(root);
  const ignored = new Set(opts.ignore ?? []);
  const findings: Finding[] = [];
  for (const rule of ALL_RULES) {
    if (ignored.has(rule.ruleId)) continue;
    findings.push(...rule.check(layout, spec));
  }
  findings.sort((a, b) => {
    if (a.ruleId !== b.ruleId) return a.ruleId < b.ruleId ? -1 : 1;
    if (a.path !== b.path) return a.path < b.path ? -1 : 1;
    return a.line - b.line;
  });
  return {
    specVersion: spec.spec_version,
    claudeAppVersion: spec.claude_app_version,
    findings,
  };
}
